### PR TITLE
(WIP) ENH: Add PeriodBlock

### LINKED ISSUE #7964: API/ENH: allow Period to be a first-class type via a PeriodBlockinternal type
Instead of having a represenation of an array of `Period` objects (object dtype) in a Series/DataFrame. Allow an internal type of `PeriodBlock` to represent, similary to `DatetimeBlock`.

Here is a usecase. The following would be very simple if `to_period(...)` returned a `PeriodIndex`

```
In [59]: df = DataFrame({'date' : pd.date_range('20130101 00:01:00',periods=5,freq='61s')})

In [60]: df
Out[60]: 
                 date
0 2013-01-01 00:01:00
1 2013-01-01 00:02:01
2 2013-01-01 00:03:02
3 2013-01-01 00:04:03
4 2013-01-01 00:05:04

In [61]: df.date.dt.to_period('T')                   
Out[61]: 
0    2013-01-01 00:01
1    2013-01-01 00:02
2    2013-01-01 00:03
3    2013-01-01 00:04
4    2013-01-01 00:05
dtype: object

In [62]: pd.Series(pd.PeriodIndex(df.date.dt.to_period('T')+1).to_timestamp())
Out[62]: 
0   2013-01-01 00:02:00
1   2013-01-01 00:03:00
2   2013-01-01 00:04:00
3   2013-01-01 00:05:00
4   2013-01-01 00:06:00
dtype: datetime64[ns]
```

### PR DESCRIPTION
**Pull request by sinhrks**
- [x] closes #7964
- [ ] tests added / passed
- [x] passes `git diff upstream/master | flake8 --diff`
- [ ] whatsnew entry

This PR is **NOT** ready for review, but to show basic direction. Especially, supporting period arithmetic is quite unreliable because current `_TimeOp` class is basically for `np.ndarray` operations. I think refactoring `_TimeOp` is needed prior to this PR.

Would like to split the PR to 3 parts:
- 1. Add `PeriodDtype` definition, let `PeriodIndex` to have `PeriodDtype` like `period[M]`.  (#13941)
- 2. Refactor `_TimeOp` splitting to `_DatetimeOp` and `_TimedeltaOp`, merging `DatetimeIndex` and `TimedeltaIndex` arithmetics.
  - `_DatetimeOp` is for lhs or rhs is `datetime64` or `datetimetz` dtypes.
  - `_TimedeltaOp` is for other `timedelta` related types.
- 3. Add `PeriodBlock`, adding `_PeriodOp` for arithmetic (this PR).

### CODE REVIEW

File: pandas/core/internals.py [hunk 1]
```diff
@@ -2451,6 +2462,182 @@ def shift(self, periods, axis=0, mgr=None):
                                            placement=self.mgr_locs)]
 
 
+class PeriodBlock(NonConsolidatableMixIn, DatetimeBlock):
```
Comment (jreback) [inline pandas/core/internals.py]: you might be better actoring out some of the holding Index suff from ``DatetimeTZBLock`` and creating ``IndexHolderBlock`` (or better name). then sub-classing for ``DatetimeTZBlock`` and ``PeriodBlock``. Should be in theory less duplication.

File: pandas/core/internals.py [hunk 2]
```diff
@@ -2451,6 +2462,182 @@ def shift(self, periods, axis=0, mgr=None):
...
+                other = other.astype('i8', copy=False).view('i8')
+            except ValueError:
+
+                # coercion issues
+                # let higher levels handle
+                raise TypeError
+
+        return values, values_mask, other, other_mask
+
+    def _try_coerce_result(self, result):
```
Comment (jreback) [inline pandas/core/internals.py]: if you factor as I suggest above, you can prob make a
```
def _shallow_copy_index(result):
    return PeriodIndex(result, freq=self.values.freq)
```

to do the new internal index creation (and same for ``DatetimeTZBlock``) to make this logic simpler (maybe)

File: pandas/core/internals.py [hunk 3]
```diff
@@ -4444,18 +4647,18 @@ def _lcd_dtype(l):
...
     have_cat = len(counts[CategoricalBlock]) > 0
     # TODO: have_sparse is not used
     have_sparse = len(counts[SparseBlock]) > 0  # noqa
     have_numeric = have_float or have_complex or have_int
 
     has_non_numeric = have_dt64 or have_dt64_tz or have_td64 or have_cat
+    has_datetime_like = have_dt64 or have_dt64_tz or have_td64 or have_period
 
     if (have_object or
-        (have_bool and
```
Comment (jreback) [inline pandas/core/internals.py]: this whole section needs factoring, but later

File: pandas/core/reshape.py [hunk 4]
```diff
@@ -187,8 +195,14 @@ def get_new_values(self):
             dtype = values.dtype
             new_values = np.empty(result_shape, dtype=dtype)
         else:
-            dtype, fill_value = _maybe_promote(values.dtype, self.fill_value)
-            new_values = np.empty(result_shape, dtype=dtype)
+            if self.is_period is not None:
+                dtype, fill_value = _maybe_promote(self.is_period.dtype,
```
Comment (jreback) [inline pandas/core/reshape.py]: this is getting awkward......

File: pandas/core/internals.py [hunk 5]
```diff
@@ -2451,6 +2464,168 @@ def shift(self, periods, axis=0, mgr=None):
...
+        if is_period_dtype(dtype):
+            dtype = PeriodDtype(dtype)
+            values = self.values
+            # ToDo: asfreq??
+            return self.make_block(values)
+
+        # delegate
+        return super(PeriodBlock, self)._astype(dtype=dtype, **kwargs)
+
+    def external_values(self):
```
Comment (jreback) [inline pandas/core/internals.py]: can move to IndexHolder

File: pandas/core/internals.py [hunk 6]
```diff
@@ -2451,6 +2464,168 @@ def shift(self, periods, axis=0, mgr=None):
...
+        """ we internally represent the data as a PeriodIndex, but for external
+        compat with ndarray, export as a ndarray of Period """
+        return self.values.asobject.values
+
+    def to_dense(self):
```
Comment (jreback) [inline pandas/core/internals.py]: same?

File: pandas/core/internals.py [hunk 7]
```diff
@@ -2451,6 +2464,168 @@ def shift(self, periods, axis=0, mgr=None):
...
+                result[mask] = tslib.iNaT
+        elif isinstance(result, np.integer):
+            result = self._box_func(result)
+        return result
+
+    def should_store(self, value):
+        return isinstance(value, Period) and self.values.freq == value.freq
+
+    def to_native_types(self, slicer=None, na_rep=None,
+                        quoting=None, **kwargs):
```
Comment (jreback) [inline pandas/core/internals.py]: can also move (if use box_func) (though this IS datetimelike specific, but ok to stick in IndexHolder)

File: pandas/core/internals.py [hunk 8]
```diff
@@ -2451,6 +2464,168 @@ def shift(self, periods, axis=0, mgr=None):
...
+        if na_rep is None:
+            na_rep = 'NaT'
+        rvalues[mask] = na_rep
+        imask = (~mask).ravel()
+
+        rvalues.flat[imask] = np.array([Period(val)._repr_base(format='all')
+                                        for val in values.ravel()[imask]],
+                                       dtype=object)
+        return rvalues
+
```
Comment (jreback) [inline pandas/core/internals.py]: same here

File: pandas/core/internals.py [hunk 9]
```diff
@@ -4444,18 +4635,18 @@ def _lcd_dtype(l):
...
     has_non_numeric = have_dt64 or have_dt64_tz or have_td64 or have_cat
+    has_datetime_like = have_dt64 or have_dt64_tz or have_td64 or have_period
 
     if (have_object or
-        (have_bool and
-         (have_numeric or have_dt64 or have_dt64_tz or have_td64)) or
-        (have_numeric and has_non_numeric) or have_cat or have_dt64 or
-            have_dt64_tz or have_td64):
+       (have_bool and (have_numeric or has_datetime_like)) or
+       (have_numeric and has_non_numeric) or have_cat or has_datetime_like):
```
Comment (jreback) [inline pandas/core/internals.py]: should prob move this entire mess to ``pandas.types.cast`` (maybe sep PR and do before?)

File: pandas/core/series.py [hunk 10]
```diff
@@ -2861,6 +2864,8 @@ def create_from_value(value, index, dtype):
 
         if is_datetimetz(dtype):
```
Comment (jreback) [inline pandas/core/series.py]: so now seems that we have tons of cases where doing:

```
if is_datetimetz: 
   ....
elif is_period.....
    ...

elif is_categorical...
    ...
```

ideally we should have functions for various things based on the dtype object itself.

e.g. something like:

```
try:
    subarr = dtype._create_array(......)
except:
    pass
```

Comment (jreback) [inline pandas/core/series.py]: you may need to make this a single routine something like in ``pandas.types.dtypes```

```
def _create_array(arr, dtype):
     if is_extension_dtype(dtype):
             # _type_class is the Index type, but not sure how this should be attached to dtype.
             return dtype._type_class(arr)
     return np.asarray(arr, dtype=dtype)

```

File: pandas/tools/merge.py [hunk 11]
```diff
@@ -1151,6 +1152,10 @@ def _factorize_keys(lk, rk, sort=True):
     if is_datetime64tz_dtype(lk) and is_datetime64tz_dtype(rk):
         lk = lk.values
         rk = rk.values
+    elif is_period_dtype(lk) and is_period_dtype(rk):
```
Comment (jreback) [inline pandas/tools/merge.py]: make an or

File: pandas/tools/util.py [hunk 12]
```diff
@@ -138,7 +138,10 @@ def to_numeric(arg, errors='raise', downcast=None):
 
     if isinstance(arg, pd.Series):
         is_series = True
-        values = arg.values
```
Comment (jreback) [inline pandas/tools/util.py]: hmm, we don't have to do this for other datetimelikes?

File: pandas/types/cast.py [hunk 13]
```diff
@@ -638,9 +651,28 @@ def _soft_convert_objects(values, datetime=True, numeric=True, timedelta=True,
     if datetime:
         values = lib.maybe_convert_objects(values, convert_datetime=datetime)
 
-    if timedelta and is_object_dtype(values.dtype):
-        # Object check to ensure only run if previous did not convert
-        values = lib.maybe_convert_objects(values, convert_timedelta=timedelta)
+    if is_object_dtype(values.dtype):
```
Comment (jreback) [inline pandas/types/cast.py]: I don't think we need to add ``period`` conversions here. It is quite difficult to know whether dates or periods are requested (yes, obviously the user is providing the flag), but in reality they should just use ``.to_datetime`` (or ``.to_period``)? maybe better just create that.

File: pandas/core/series.py (hunk 10 above)
Comment (sinhrks) [inline pandas/core/series.py]: yeah, it may cover #12758 also. Let me try...

File: pandas/core/reshape.py [hunk 14]
```diff
@@ -68,10 +68,14 @@ def __init__(self, values, index, level=-1, value_columns=None,
...
 
         self.is_categorical = None
+        self.is_period = None
         if values.ndim == 1:
             if isinstance(values, Categorical):
                 self.is_categorical = values
                 values = np.array(values)
+            elif is_period(values):
+                self.is_period = values
+                values = values.values
```
Comment (jreback) [inline pandas/core/reshape.py]: I guess not handling other i8 here ATM? (e.g. datetime/datetimetz)?

File: pandas/core/series.py [hunk 15]
```diff
@@ -161,8 +162,9 @@ def __init__(self, data=None, index=None, dtype=None, name=None,
                 # need to copy to avoid aliasing issues
                 if name is None:
                     name = data.name
-
-                data = data._to_embed(keep_tz=True)
+                if not isinstance(data, ABCPeriodIndex):
+                    # to_embed returns Period object array
```
Comment (jreback) [inline pandas/core/series.py]: should do this generically based on index type

Comment (jreback) [inline pandas/core/series.py]: this ``_to_embed`` overlaps with ``_sanitize_index``

File: pandas/core/series.py [hunk 16]
```diff
@@ -2797,8 +2798,12 @@ def _try_cast(arr, take_fast_path):
             if not is_extension_type(subarr):
                 subarr = np.array(subarr, dtype=dtype, copy=copy)
         except (ValueError, TypeError):
+
             if is_categorical_dtype(dtype):
                 subarr = Categorical(arr)
+            elif is_period_dtype(dtype):
```
Comment (jreback) [inline pandas/core/series.py]: should these be handled in ``_possibly_cast_to_datetime``?

File: pandas/indexes/base.py [hunk 17]
```diff
@@ -162,20 +165,25 @@ def __new__(cls, data=None, dtype=None, copy=False, name=None,
         # index-like
         elif isinstance(data, (np.ndarray, Index, ABCSeries)):
 
-            if (issubclass(data.dtype.type, np.datetime64) or
-                    is_datetimetz(data)):
+            if is_datetime64_any_dtype(data):
                 from pandas.tseries.index import DatetimeIndex
                 result = DatetimeIndex(data, copy=copy, name=name, **kwargs)
```
Comment (jreback) [inline pandas/indexes/base.py]: maybe make a function to do this:

```
if is_datetime64_any_dtype(data):
     from pandas.tseries.index import DatetimeIndex
     return DatetimeIndex._from_index(data, copy=copy, name=name, **kwargs)
....
# timedelta
# period
```

IOW make the coercion happen inside the ``_from_index`` class constructor

**Comment by jreback** [comment]
@sinhrks looks pretty good. certainly can merge mostly as is and just make a to-do on other issues. Looks like pretty much works and doesn't change the api much.

Just on a small whatsnew section to we can keep updating.

**Comment by jreback** [comment]
frrom #13345 

```
In [31]: dates = pd.date_range('1/1/2011', periods=500000, freq='H')
    ...: values = np.random.random(size=len(dates))
    ...: df = pd.DataFrame({"Date":dates, "Values":values})
    ...: 

In [32]: df
Out[32]: 
                      Date    Values
0      2011-01-01 00:00:00  0.787923
1      2011-01-01 01:00:00  0.730256
2      2011-01-01 02:00:00  0.032293
3      2011-01-01 03:00:00  0.411157
4      2011-01-01 04:00:00  0.295745
5      2011-01-01 05:00:00  0.815852
...                    ...       ...
499994 2068-01-15 02:00:00  0.368692
499995 2068-01-15 03:00:00  0.676244
499996 2068-01-15 04:00:00  0.123631
499997 2068-01-15 05:00:00  0.640574
499998 2068-01-15 06:00:00  0.709926
499999 2068-01-15 07:00:00  0.665387

[500000 rows x 2 columns]

In [33]: df.dtypes
Out[33]: 
Date      datetime64[ns]
Values           float64
dtype: object

In [34]: df['period'] = df.Date.dt.to_period(freq="1M")

In [35]: df.head()
IndexError: too many indices for array
```

**Comment by jreback** [comment]
This should return a Series
```
In [61]: x
Out[61]: 
0        2011-01
1        2011-01
2        2011-01
3        2011-01
4        2011-01
5        2011-01
           ...  
499994   2068-01
499995   2068-01
499996   2068-01
499997   2068-01
499998   2068-01
499999   2068-01
Name: Date, dtype: period[M]

In [62]: x.dt.to_timestamp()
Out[62]: 
DatetimeIndex(['2011-01-01', '2011-01-01', '2011-01-01', '2011-01-01', '2011-01-01', '2011-01-01', '2011-01-01', '2011-01-01', '2011-01-01', '2011-01-01',
               ...
               '2068-01-01', '2068-01-01', '2068-01-01', '2068-01-01', '2068-01-01', '2068-01-01', '2068-01-01', '2068-01-01', '2068-01-01', '2068-01-01'],
              dtype='datetime64[ns]', name=u'Date', length=500000, freq=None)
```

This is correct (the reverse)
```
In [64]: s2 = Series(x.dt.to_timestamp())

In [65]: s2.dt.to_period('M')
Out[65]: 
0        2011-01
1        2011-01
2        2011-01
3        2011-01
4        2011-01
5        2011-01
           ...  
499994   2068-01
499995   2068-01
499996   2068-01
499997   2068-01
499998   2068-01
499999   2068-01
Name: Date, dtype: period[M]
```

**Comment by jreback** [comment]
something going on here (prob to/from object conversion)

```
In [71]: s2
Out[71]: 
0        2011-01-01
1        2011-01-01
2        2011-01-01
3        2011-01-01
4        2011-01-01
5        2011-01-01
            ...    
499994   2068-01-01
499995   2068-01-01
499996   2068-01-01
499997   2068-01-01
499998   2068-01-01
499999   2068-01-01
Name: Date, dtype: datetime64[ns]

In [72]: x
Out[72]: 
0        2011-01
1        2011-01
2        2011-01
3        2011-01
4        2011-01
5        2011-01
           ...  
499994   2068-01
499995   2068-01
499996   2068-01
499997   2068-01
499998   2068-01
499999   2068-01
Name: Date, dtype: period[M]
```

```
In [69]: %timeit s2[s2>'2015']
100 loops, best of 3: 2.66 ms per loop

In [70]: %timeit x[x>'2015']
1 loop, best of 3: 637 ms per loop
```

**Comment by jorisvandenbossche** [comment]
@sinhrks @jreback What is the status of this one? Still try to include it in 0.19.0 or push to 0.20?

**Comment by jreback** [comment]
ideally I'd like this to go in. I think most of the prep work has been done.

**Comment by jorisvandenbossche** [comment]
@sinhrks Is this ready for review? (no docs yet, and some todo's listed in the code) Just to prioritize my time :-)

**Comment by sinhrks** [comment]
Not yet. I'll make a notice once ready, hopefully this weekend... :)

File: pandas/core/reshape.py [hunk 18]
```diff
@@ -69,10 +69,14 @@ def __init__(self, values, index, level=-1, value_columns=None,
                  fill_value=None):
 
         self.is_categorical = None
+        self.is_period = None
         if values.ndim == 1:
             if isinstance(values, Categorical):
```
Comment (jreback) [inline pandas/core/reshape.py]: -> is_categorical

File: pandas/io/tests/test_feather.py [hunk 19]
```diff
@@ -84,7 +84,7 @@ def test_unsupported(self):
 
         # period
         df = pd.DataFrame({'a': pd.period_range('2013', freq='M', periods=3)})
-        self.check_error_on_write(df, ValueError)
+        self.check_error_on_write(df, feather.FeatherError)
```
Comment (jreback) [inline pandas/io/tests/test_feather.py]: yeah I don't think feather supports Periods ATM.

File: pandas/src/hashtable_class_helper.pxi.in [hunk 20]
```diff
@@ -488,6 +506,8 @@ cdef class StringHashTable(HashTable):
             if k != self.table.n_buckets:
                 val = self.table.vals[k]
 
+    @cython.wraparound(False)
```
Comment (jreback) [inline pandas/src/hashtable_class_helper.pxi.in]: this file changes can do as a separate PR?

File: pandas/tseries/frequencies.py [hunk 21]
```diff
@@ -773,7 +772,7 @@ def infer_freq(index, warn=True):
                             "dtype on a Series of {0}".format(index.dtype))
         index = values
 
-    if is_period_arraylike(index):
+    if isinstance(index, pd.PeriodIndex):
```
Comment (jreback) [inline pandas/tseries/frequencies.py]: is_period_dtype

File: pandas/types/cast.py [hunk 22]
```diff
@@ -265,6 +267,12 @@ def _maybe_promote(dtype, fill_value=np.nan):
     elif is_datetimetz(dtype):
         if isnull(fill_value):
             fill_value = iNaT
+    elif is_period(dtype):
+        from pandas.tseries.period import Period
+        if isnull(fill_value):
+            fill_value = tslib.iNaT
+        elif isinstance(fill_value, Period):
```
Comment (jreback) [inline pandas/types/cast.py]: you can use ABCPeriod here

File: pandas/types/common.py [hunk 23]
```diff
@@ -73,7 +73,7 @@ def is_datetimetz(array):
 
 def is_period(array):
     """ return if we are a period array """
-    return isinstance(array, ABCPeriodIndex) or is_period_arraylike(array)
```
Comment (jreback) [inline pandas/types/common.py]: hmm so this is now the same as ``is_period_dtype``, maybe we should just deprecate this then, unless this is *really* equiv to ``isinstance(val, Period)`` ? IOW a scalar
